### ISSUE TITLE
Spread ABM scheduling across multiple globalsteps without lagspikes

### ISSUE THREAD

**Issue by Warr1024** [minetest/minetest#14011]
### Problem

ABM scheduling right now is a "worst of both worlds" situation:

- With a 20% (default) time budget in place, ABMs are prevented from accessing 80% of CPU time, even if that time is not actually used by anything else.
- With the default 20% time budget and an interval of 1s, ABMs can still create a 200ms lagspike every second, which can lead to notable performance judder on servers with dedicated_step < 0.2 (virtually all of them).

This is because the ABM timer runs only once every second, and attempts to use 20% of that 1 second interval to run ABMs all at once, rather than spreading ABM runs out across server steps.

### Solutions

- When the ABM timer fires, add all loaded mapblocks to a queue that are not yet in there.
- Each server step (possibly after processing most other things for that step), start dequeueing mapblocks in random order from the queue, and then process runnable ABMs for each one.
- Before processing a mapblock, make sure it's still loaded and active; if not, then just skip it and move on to the next.
- Stop processing mapblocks when either the queue is empty, or the dedicated_step time budget for that step has been exceeded.

We can then get rid of the 20% time budget hack.  We can offer additional settings, if desired, to limit CPU usage (an alternative budget, or a time margin, after which we will prefer to sleep rather than dequeueing more blocks).

### Alternatives

Another alternative is just to make the basic machinery for ABMs available to the Lua API and let game/mod devs manage their own scheduling:

- API method to get the list the active mapblocks for the current globalstep.
- API method to do an "ABM scan", finding all nodes matching nodenames and meeting the neighbor criteria within a mapblock.

This could be a little more complex, but would empower modders to solve this problem to their own satisfaction while also absolving the engine team from having to deal with arbitrarily complex scheduling logic.

### Additional context

The last major work (that I know of) done on the ABM scheduler was in https://github.com/minetest/minetest/pull/10290, which was in response to https://github.com/minetest/minetest/issues/10068.  It looked like it would have allowed us to increase the ABM time budget and run more ABMs with available CPU, but in practice it was not actually possible to do that without increasing server lag.

**Comment by Warr1024** [comment]
N.B. general guidance for ABM performance issues has been "use node timers instead", but I ran into multiple issues with those:

- They are unreliable (some things, like apparently having too many node timers active per mapblock) can cause node timers to be lost.  LBMs can only regenerate them if the mapblock is freshly loaded (and there's nothing players can do independently to force this, not that they should ever have to) so you end up needing an ABM to patrol and regenerate them anyway.
- You cannot register multiple node timers for a single node, e.g. in a dependent mod, the way you can with ABMs.  I ended up hacking in a whole system for this into NodeCore, but it makes the game incompatible with some other mods (since the base node timer behavior had to be blocked since there's no way to automatically up-convert it).  This is a significant gap between the ABM and Node timer APIs.
- The fact that ABMs are not reliable, and that they are lost to preserve CPU in other areas, is actually _desirable_.  I would like to be able to ask a lot from ABMs and just get a proportional fraction of that up to the capability of the system.  I just don't want to get a 200ms lag spike while still leaving 80% of available performance on the table.

**Comment by jeremyshannon** [comment]
Yeah on Exile's v4 development branch we've had to throw ABMs out the window for most tasks due to their various problems. (kromka-chieba has written a "mapchunk shepherd" that handles our seasonal changes and water movement using voxelmanips) We still have a few things that need ABMs though, so improvements there would be very welcome!

**Comment by Warr1024** [comment]
I don't see how changing the ABM timer interval would help when ABMs are defined by mods/games, and historically all use integer multiples, so I would still end up with the Thundering Herd Effect on those integer-aligned ticks.

If I set an 80% time budget and make ABMs check every 0.1s tick, then doesn't that mean that each ABM run would just be exhausted after 80ms instead of 200ms?  That just sounds like fewer mapblocks processed each time the ABM fires.

If I have e.g. an ABM with interval 1 and chance 1, then there's nothing I can do to spread it out fairly across more than 1 tick each second.  I can't use, say, interval 0.1 and chance 10, because which nodes are run each pass will be independent, so some may fire many as 10x in one second, and others 0x; even if the average is 1x, it's still very far from the original behavior.

Also, having ABMs run their full process more often instead of queueing the results of past runs also discards more work, so while I may be able to use more CPU in running ABMs, it would mostly just go to increased overhead.  Most of my ABMs in practice are mostly overhead as it is; the most costly things are just scanning for nodes and doing things like neighbor checks.

I believe I experimented with settings like you described around the time the scheduler was first changed, and discovered that they didn't actually help and that the sanest settings were just the defaults anyway.  There's only so much that fine tuning can help to compensate for flawed assumptions.

**Comment by Warr1024** [comment]
If you're talking about how I would use the settings I'm proposing, I'm thinking like setting dedicated_step to 0.1s and ABMs per step to like 0.05s.  That will get me 2.5x as much total time for ABMs while cutting my max lag in half.

I can't do that with the settings we have now though, because if I set ABM timer to 0.1s and budget to 50%, then I will only get 1 tick of 0.05s running ABMs per second for a 1 second interval ABM, instead of the 0.5s I should be getting across all 10 ticks.

**Comment by ancientmarinerdev** [comment]
Just to add that the moving things to node timers is pretty abysmal, especially for common blocks. On MineClone2, I moved from node timers to ABM on kelp, which historically was moved from ABM to node timers because performance was bad, but it was much worse on node timers:

Went from using 11.64% CPU to 2.70%. Profiling data is attached here:

https://git.minetest.land/MineClone2/MineClone2/pulls/3417

**Comment by jmole** [comment]
QQ - does the status quo prevent ABMs from running at more than 5Hz?

I've been trying to figure out why my NodeTimerRef won't get a callback sooner than 200ms. e.g something like this:
```lua
core.get_node_timer(pos):start(0.1)

-- inside node def
on_timer = function(pos, elapsed)
  core.get_node_timer(pos):start(0.1)
  DEBUG(elapsed)
  -- etc
end
```
Just prints: 
- 0.2
- 0.2
- 0.2
- ...

**Comment by sfence** [comment]
@jmole 
Your server step is probably set to 200 ms.

**Comment by sfan5** [comment]
It's this setting:
https://github.com/luanti-org/luanti/blob/47c75b3294fc01250488cf7304dd786b8e120e4e/builtin/settingtypes.txt#L2135-L2136
200ms being the default
